File: src/main/java/dev/jlo/gamemodes/domain/common/DefaultMatchCoordinator.java
package dev.jlo.gamemodes.domain.common;

import java.time.Clock;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public class DefaultMatchCoordinator implements MatchCoordinator {
    private final Clock clock;
    private final Map<MatchId, Match> matches = new LinkedHashMap<>();
    private final PlayerOwnership ownership = new PlayerOwnership();

    public DefaultMatchCoordinator() {
        this(Clock.systemUTC());
    }

    public DefaultMatchCoordinator(Clock clock) {
        this.clock = clock;
    }

    @Override
    public synchronized void register(Match match) {
        if (matches.containsKey(match.getId())) {
            throw new IllegalStateException("Match is already registered: " + match.getId().getValue());
        }
        matches.put(match.getId(), match);
    }

    @Override
    public synchronized void unregister(MatchId id) {
        matches.remove(id);
        ownership.releaseOwner(new Ownership.Match(id.getValue()));
    }

    @Override
    public synchronized Ownership ownerOf(UUID player) {
        return ownership.ownerOf(player, clock.instant());
    }

    public synchronized void claim(UUID player, Ownership owner) {
        ownership.claim(player, owner);
    }

    public synchronized boolean leave(UUID player) {
        return ownership.releaseQueue(player);
    }

    public synchronized void disconnect(UUID player, Duration reservationDuration) {
        ownership.disconnect(player, clock.instant(), reservationDuration);
    }

    public synchronized List<MatchEvent> dispatch(MatchId id, MatchEvent event) {
        Match match = matches.get(id);
        if (match == null) {
            throw new IllegalStateException("Unknown match: " + id.getValue());
        }
        return match.handle(event);
    }
}

File: src/main/kotlin/dev/jlo/gamemodes/domain/common/DefaultMatchCoordinator.kt
package dev.jlo.gamemodes.domain.common

import java.time.Clock
import java.time.Duration
import java.util.UUID

class DefaultMatchCoordinator(
    private val clock: Clock = Clock.systemUTC()
) : MatchCoordinator {
    private val matches = linkedMapOf<MatchId, Match>()
    private val ownership = PlayerOwnership()

    @Synchronized
    override fun register(match: Match) {
        check(match.id !in matches) { "Match is already registered: ${match.id.value}" }
        matches[match.id] = match
    }

    @Synchronized
    override fun unregister(id: MatchId) {
        matches.remove(id)
        ownership.releaseOwner(Ownership.Match(id.value))
    }

    @Synchronized
    override fun ownerOf(player: UUID): Ownership? = ownership.ownerOf(player, clock.instant())

    @Synchronized
    fun claim(player: UUID, owner: Ownership) = ownership.claim(player, owner)

    @Synchronized
    fun leave(player: UUID): Boolean = ownership.releaseQueue(player)

    @Synchronized
    fun disconnect(player: UUID, reservationDuration: Duration) {
        ownership.disconnect(player, clock.instant(), reservationDuration)
    }

    @Synchronized
    fun dispatch(id: MatchId, event: MatchEvent): List<MatchEvent> = matches[id]?.handle(event)
        ?: error("Unknown match: ${id.value}")
}
